### PR TITLE
More flexible ways to configure TracerProviderBuilder when using Ext.Hosting package.

### PR DESCRIPTION
**Pull request by cijothomas**
The PR also modified the example asp.net core app, to show the flexibility enabled by this PR. Some parts look a bit ugly. Need to see if there is easier alternates...

This can address https://github.com/open-telemetry/opentelemetry-dotnet/issues/831, https://github.com/open-telemetry/opentelemetry-dotnet/issues/1394 once refined fully :)

## Changes

Please provide a brief description of the changes here.

For significant contributions please make sure you have completed the following items:

* [ ] `CHANGELOG.md` updated for non-trivial changes
* [ ] Design discussion issue #
* [ ] Changes in public API reviewed

### CODE REVIEW

File: examples/AspNetCore/Startup.cs [hunk 1]
```diff
@@ -92,15 +93,33 @@ public void ConfigureServices(IServiceCollection services)
                         }));
                     break;
                 default:
-                    services.AddOpenTelemetryTracing((builder) => builder
-                        .AddAspNetCoreInstrumentation()
-                        .AddHttpClientInstrumentation()
-                        .AddConsoleExporter());
+                    // Adding OtelTracing. Prepares to accept more configs from DI.
```
Comment (cijothomas) [inline examples/AspNetCore/Startup.cs:96]: @tekian Please take a look. (In context of https://github.com/open-telemetry/opentelemetry-dotnet/issues/1394)

File: examples/AspNetCore/Startup.cs [hunk 2]
```diff
@@ -109,6 +128,18 @@ public void ConfigureServices(IServiceCollection services)
                         };
                     });
 
+                    // Configure builder, but also need to get hold of IServiceProvider.
```
Comment (cijothomas) [inline examples/AspNetCore/Startup.cs:131]: @alanwest 
From this comment: https://github.com/open-telemetry/opentelemetry-dotnet/pull/1469/files#r518372009

File: src/OpenTelemetry.Extensions.Hosting/Implementation/TracerProviderBuilderHosting.cs [hunk 3]
```diff
@@ -83,6 +84,12 @@ public TracerProviderBuilder Configure(Action<IServiceProvider, TracerProviderBu
 
         public TracerProvider Build(IServiceProvider serviceProvider)
         {
+            var configActions = serviceProvider.GetRequiredService<IEnumerable<IConfigureOptions<TracerProviderBuilder>>>();
```
Comment (CodeBlanch) [inline src/OpenTelemetry.Extensions.Hosting/Implementation/TracerProviderBuilderHosting.cs:87]: This is either totally brilliant or misuse of the options API 😄 I suppose `TracerProviderBuilder` is kind of an options class? I think the intention with `IConfigureOptions` is the registrations are invoked on options classes created when fed into something via IOptions<T> (or its cousins). This is kind of co-opting that mechanism for our own purpose, which isn't necessarily a bad thing. We could have our own `IConfigureTracerProviderBuilder` that does essentially the same thing without the co-opt.

**Comment by CodeBlanch** [comment]
I'm not totally understanding the use-case for this yet. For library authors wanting to help users configure their TracerProviders, I don't think calling `services.Configure` is the way to go about it. I feel like the standard pattern in .NET is to do that via extensions on the builder, a la:

MyLibrary:
```csharp
public static void UseMyExtension(this TracerProviderBuilder builder) { ... }
```

User's Startup.cs:
```csharp
                services.AddOpenTelemetryTracing(builder =>
                {
                    builder.UseMyExtension();
                });

                services.AddOpenTelemetryTracing(builder =>
                {  
                    // Configured to send http spans to AppInsights
                });

                services.AddOpenTelemetryTracing(builder =>
                {                   
                    // Configured to write everything to stdout
                });
```

I don't see what is wrong with the above. I think it looks _great_! 😄 Really easy for users to do the single and multiple provider cases.

This PR (I think) is trying to achieve something like this...

User's Startup.cs:
```csharp
                services.AddOpenTelemetryTracing(builder =>
                {
                    // I want to do something custom hee
                });

                services.AddOpenTelemetryTracing(builder =>
                {  
                    // Configured to send http spans to AppInsights
                });

                services.AddOpenTelemetryTracing(builder =>
                {                   
                    // Configured to write everything to stdout
                });

               services.AddMyCustomOpenTelemetryExtension(); // Magic happens to apply this to the correct TracerProvider
```

It kind of breaks the builder pattern.

Sure we could disallow multiple calls to `AddOpenTelemetryTracing`, but why? It looks and works _great_ this way (with #2004) and is super easy for users. 💯 🥇 🚀 

To enable what this PR is trying to do, the only thing I've come up with so far is to add named TracerProviders a la...

User's Startup.cs:
```csharp
                services.AddOpenTelemetryTracing("Default", builder =>
                {
                    // I want to do something custom hee
                });

                services.AddOpenTelemetryTracing("AppInsights", builder =>
                {  
                    // Configured to send http spans to AppInsights
                });

                services.AddOpenTelemetryTracing("stdout", builder =>
                {                   
                    // Configured to write everything to stdout
                });

               services.AddMyCustomOpenTelemetryExtension(applyToTracerProviderNames: "Default", "stdout");
               services.AddSomeOtherCustomOpenTelemetryExtension(); // Does this apply to just Default or all providers?
```

I don't think supporting that would be too hard. It could be non-breaking, on `AddOpenTelemetryTracing` we just add a param `string providerName = "Default"`. Users using one provider could just use the default name forever 🤷

**Comment by alanwest** [comment]
I'm trying to get my head around the use cases for multiple calls to `AddOpenTelemetryTracing`.

@CodeBlanch's code snippets suggests a use case for enabling the configuration of multiple trace providers differently (e.g., each may use a different means of export and different instrumentation).

Though #1394 seems to suggest a desire for a different use case, namely the ability to call `AddOpenTelemetryTracing` from multiple locations where each call builds upon on the configuration from previous calls.

If I'm understanding things correctly, I think these two use cases are at odds with one another.

**Comment by CodeBlanch** [comment]
@alanwest Yes nail on the head. In my mind, there are 3 challenges:

1) We need to support multiple tracer providers. In the .NET Framework world, how you do this is obvious... just new up a builder for each provider you need. But in .NET Core world, it isn't as obvious because there is this DI pattern. Talking with @cijothomas, he is OK with users only being able to use the extension to configure a single provider. If they want more than one, they can use the .NET Framework pattern (basically, do it manually with little or no DI). What I'm trying to show on #2006 and on #2004 is that we are really close to supporting multiple providers through the extension, and I think this looks really clean and is very user-friendly.

2) The second thing is allowing users to register globally a Sampler or Processor that the SDK will automatically pick up without doing anything to the builder. That is what #2006 is doing. From what @cijothomas tells me, AppInsights works like this so customers kind of expect it.

3) The third thing is what is on here. A way for any random library to register into DI and take over the configuration of tracer provider in some way. This is at odds with the first goal. If there are multiple tracer providers, how does a library know what to do? Could be done with named tracer providers?

**Comment by aelij** [comment]
I think a cleaner (and more MS.Extensions-conventional) solution would be to return a builder interface that's purely based on DI. This way the builder instance has no state, so it's irrelevant how many of it the user creates. Actual state can be stored in an options class. It allows the user to fully customize the flow as all the actions on the builder are basically just DI shortcuts.

For example (incomplete):

```cs
ITraceBuilder AddOpenTelemetryTraces(this IServiceCollection services) {
   services.AddSingleton(sp => new TracerProviderBuilder(sp.GetRequiredService<IOptions<TraceBuilderOptions>(),
      sp.GetServices<InstrumentationFactory>(), ...).Build());
   return new TraceBuilder(services);
}

class TraceBuilder : ITraceBuilder
{
   IServiceProvider Services {get;}

   ITraceBuilder AddInstrumentation<TInstrumentation>() {
    Services.TryAddEnumerable(ServiceDescriptor.Singleton(new InstrumentationFactory(...));
    return this;
  }

  ITraceBuilder AddSource(string s) {
    Services.Configure<TraceBuilderOptions>(o => o.Sources.Add(s));
  }
}

**Pull request closed without merging**